libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioAdapter.cs

```diff
@@ -87,71 +88,46 @@ public override async Task<ResourceResponse[]> SendActivitiesAsync(ITurnContext
-
-            var twilioSignature = request.Headers["x-twilio-signature"];
-
-            var validationUrl = _options.ValidationUrl ?? (request.Headers["x-forwarded-proto"][0] ?? request.Protocol + "://" + request.Host + request.Path);
+            if (httpRequest == null)
+            {
+                throw new ArgumentNullException(nameof(httpRequest));
+            }
 
-            var requestValidator = new RequestValidator(_options.AuthToken);
+            if (httpResponse == null)
+            {
+                throw new ArgumentNullException(nameof(httpResponse));
+            }
 
-            using (var bodyStream = new StreamReader(request.Body))
+            if (bot == null)
             {
-                var payload = bodyStream.ReadToEnd();
+                throw new ArgumentNullException(nameof(bot));
+            }
 
-                var formattedPayload = payload.Replace("+", "%20");
+            var activity = ReadRequest(httpRequest);
 
-                var body = QueryStringToDictionary(formattedPayload);
+            // create a conversation reference
+            using (var context = new TurnContext(this, activity))
+            {
+                context.TurnState.Add("httpStatus", "200");
 
-                if (!string.IsNullOrWhiteSpace(twilioSignature) &&
-                    requestValidator.Validate(validationUrl, body, twilioSignature))
-                {
-                    var jsonBody = JsonConvert.SerializeObject(body);
+                await RunPipelineAsync(context, bot.OnTurnAsync, cancellationToken).ConfigureAwait(false);
 
-                    var twilioEvent = JsonConvert.DeserializeObject<TwilioEvent>(jsonBody);
+                httpResponse.StatusCode = Convert.ToInt32(context.TurnState.Get<string>("httpStatus"));
 
-                    if (int.TryParse(twilioEvent.NumMedia, out var numMediaResult) && numMediaResult > 0)
-                    {
-                        // specify a different event type
-                        twilioEvent.EventType = "picture_message";
-                    }
+                httpResponse.ContentType = "text/plain";
 
-                    var activity = new Activity()
-                    {
-                        Id = twilioEvent.MessageSid,
-                        Timestamp = new DateTime(),
-                        ChannelId = "twilio-sms",
-                        Conversation = new ConversationAccount() { Id = twilioEvent.From, },
-                        From = new ChannelAccount() { Id = twilioEvent.From, },
-                        Recipient = new ChannelAccount() { Id = twilioEvent.To, },
-                        Text = twilioEvent.Body,
-                        ChannelData = twilioEvent,
-                        Type = ActivityTypes.Message,
-                    };
+                var text = (context.TurnState.Get<object>("httpBody") != null) ? context.TurnState.Get<object>("httpBody").ToString() : string.Empty;
```

Comment:
Can we replace this with something like this:
```
var text = context.TurnState.Get<object>("httpBody") ?? string.Empty;
```
(In this case, you should use text_.ToString()_ in _httpResponse.WriteAsync_)

Reply:
BTW, It's possible to retrieve the _httpBody_ as a _string_?

Reply:
> Can we replace this with something like this:
> 
> ```
> var text = context.TurnState.Get<object>("httpBody") ?? string.Empty;
> ```
> 
> (In this case, you should use text_.ToString()_ in _httpResponse.WriteAsync_)

@GasparAcevedoZainSouthworks The `?? string.Empty` is not necessary as the left-hand operation can not be null. I will apply this change.